### PR TITLE
Update Downloader utility class to use static dask cluster

### LINKED ISSUE #1146: [BUG]: DFP module pipelines creating dask cluster for every batch
### Version

23.11

### Which installation method(s) does this occur on?

Docker

### Describe the bug.

Running DFP module pipelines with `DEBUG` log level shows that a new dask cluster is being created for every batch. This results in very slow processing and pipeline mostly fails before completion.

### Minimum reproducible example

1. Follow instructions [here](https://github.com/nv-morpheus/Morpheus/tree/branch-23.11/examples/digital_fingerprinting/production) to build DFP container.

2. Create `bash` shell in container:
```
docker compose run morpheus_pipeline bash
```

3. Download Azure example data:
```
python /workspace/examples/digital_fingerprinting/fetch_example_data.py all
```

4. Run module implementation of Azure DFP training pipeline using `DEBUG` log level:
```
cd /workspace/examples/digital_fingerprinting/production/morpheus
```

```
python ./dfp_integrated_training_batch_pipeline.py --use_cpp=true --source duo --start_time "2022-08-01" --duration "60d" --train_users generic --log_level DEBUG --input_file "./control_messages/duo_streaming_training.json"
```

### Relevant log output

Log snippet showing dask cluster being created multiple times:
```shell
FileToDF [training_pipe]: 915 messages [02:54,  5.23 messages/s]Creating dask cluster...
Preprocessed [training_pipe]: 1101 messages [02:53,  8.22 messagesElapsed time since last train is too short
FileToDF [training_pipe]: 915 messages [03:10,  5.23 messag/opt/conda/envs/morpheus/lib/python3.10/site-packages/merlin/dtypes/mappings/tf.py:52: UserWarning: Tensorflow dtype mappings did not load successfully due to an error: No module named 'tensorflow'
  warn(f"Tensorflow dtype mappings did not load successfully due to an error: {exc.msg}")
/opt/conda/envs/morpheus/lib/python3.10/site-packages/merlin/dtypes/mappings/tf.py:52: UserWarning: Tensorflow dtype mappings did not load successfully due to an error: No module named 'tensorflow'
  warn(f"Tensorflow dtype mappings did not load successfully due to an error: {exc.msg}")
FileToDF [training_pipe]: 915 messages [03:15,  5.23 messag2023-08-23 16:09:31,028 - distributed.preloading - INFO - Creating preload: dask_cuda.initialize
2023-08-23 16:09:31,029 - distributed.preloading - INFO - Import preload module: dask_cuda.initialize
2023-08-23 16:09:31,048 - distributed.preloading - INFO - Creating preload: dask_cuda.initialize
2023-08-23 16:09:31,048 - distributed.preloading - INFO - Import preload module: dask_cuda.initialize
FileToDF [training_pipe]: 915 messages [03:16,  5.23 messagCreating dask cluster... Done. Dashboard: http://127.0.0.1:40219/status
FileToDF [training_pipe]: 915 messages [03:17,  5.23 messagS3 objects to DF complete. Rows: 103, Cache: miss, Duration: 23103.997707366943 ms, Rate: 4.458102935456811 rows/s
FileToDF [training_pipe]: 1018 messages [03:17,  5.12 messages/s]Creating dask cluster...
Preprocessed [training_pipe]: 1101 messages [03:17,  8.22 messages/Elapsed time since last train is too short
```
Errors when pipeline fails:
```
FileToDF [training_pipe]: 1287 messages [06:06,  3.77 me2023-08-22 16:16:21,024 - distributed.worker - ERROR - Failed to communicate with scheduler during heartbeat.
Traceback (most recent call last):essages [06:06,  3.77 messages/s]s]
  File "/opt/conda/envs/morpheus/lib/python3.10/site-packages/tornado/iostream.py", line 861, in _read_to_buffer
    bytes_read = self.read_from_fd(buf)ages [06:06,  8.32 messages/s]
  File "/opt/conda/envs/morpheus/lib/python3.10/site-packages/tornado/iostream.py", line 1116, in read_from_fd
    return self.socket.recv_into(buf, len(buf)) messages/s]s/s]
ConnectionResetError: [Errno 104] Connection reset by peer/s]
Preprocessed [inference_pipe]: 0 messages [06:06, ? messages/s]
The above exception was the direct cause of the following exception:
Saved [inference_pipe]: 0 messages [06:06, ? messages/s]
```
```
Traceback (most recent call last):
  File "/opt/conda/envs/morpheus/lib/python3.10/site-packages/distributed/worker.py", line 1244, in heartbeat
    response = await retry_operation(
  File "/opt/conda/envs/morpheus/lib/python3.10/site-packages/distributed/utils_comm.py", line 434, in retry_operation
    return await retry(
  File "/opt/conda/envs/morpheus/lib/python3.10/site-packages/distributed/utils_comm.py", line 413, in retry
    return await coro()
  File "/opt/conda/envs/morpheus/lib/python3.10/site-packages/distributed/core.py", line 1265, in send_recv_from_rpc
    return await send_recv(comm=comm, op=key, **kwargs)
  File "/opt/conda/envs/morpheus/lib/python3.10/site-packages/distributed/core.py", line 1024, in send_recv
    response = await comm.read(deserializers=deserializers)
  File "/opt/conda/envs/morpheus/lib/python3.10/site-packages/distributed/comm/tcp.py", line 241, in read
    convert_stream_closed_error(self, e)
  File "/opt/conda/envs/morpheus/lib/python3.10/site-packages/distributed/comm/tcp.py", line 142, in convert_stream_closed_error
    raise CommClosedError(f"in {obj}: {exc.__class__.__name__}: {exc}") from exc
distributed.comm.core.CommClosedError: in <TCP (closed) ConnectionPool.heartbeat_worker local=tcp://127.0.0.1:52758 remote=tcp://127.0.0.1:41399>: ConnectionResetError: [Errno 104] Connection reset by peer
2023-08-22 16:16:21,096 - distributed.worker - ERROR - Failed to communicate with scheduler during heartbeat.
Traceback (most recent call last):
  File "/opt/conda/envs/morpheus/lib/python3.10/site-packages/tornado/iostream.py", line 861, in _read_to_buffer
    bytes_read = self.read_from_fd(buf)
  File "/opt/conda/envs/morpheus/lib/python3.10/site-packages/tornado/iostream.py", line 1116, in read_from_fd
    return self.socket.recv_into(buf, len(buf))
ConnectionResetError: [Errno 104] Connection reset by peer
```

### Full env printout

_No response_

### Other/Misc.

_No response_

### Code of Conduct

- [X] I agree to follow Morpheus' Code of Conduct
- [X] I have searched the [open bugs](https://github.com/nv-morpheus/Morpheus/issues?q=is%3Aopen+is%3Aissue+label%3Abug) and have found no duplicates for this bug report

### PR DESCRIPTION
**Pull request by efajardo-nv**
## Description
- `file_to_df_loader` is creating a new downloader object for every file list batch. Each downloader creates its own dask cluster which results in very slow processing and DFP module pipelines eventually fail.
- Update Downloader utility class to use static dask cluster

Fixes #1146 

## Checklist
[x] I am familiar with the [Contributing Guidelines](https://github.com/nv-morpheus/Morpheus/blob/main/docs/source/developer_guide/contributing.md).
[x] New or existing tests cover these changes.
[x] The documentation is up to date with these changes.

### CODE REVIEW

File: morpheus/loaders/file_to_df_loader.py [hunk 1]
```diff
@@ -88,7 +88,8 @@ def file_to_df_loader(control_message: ControlMessage, task: dict):
     parser_kwargs = config.get("parser_kwargs", None)
     cache_dir = config.get("cache_dir", None)
 
-    downloader = Downloader()
+    if not hasattr(file_to_df_loader, "downloader"):
+        file_to_df_loader.downloader = Downloader()
```
Comment (cwharris) [inline morpheus/loaders/file_to_df_loader.py:92]: I'm confused. Who is calling `file_to_df_loader` and why are they not responsible for making sure there is only one instance?

Comment (mdemoret-nv) [inline morpheus/loaders/file_to_df_loader.py:92]: I think it would be better to make the dask cluster static on the `Downloader` class which will avoid a possible race condition here if this is run from multiple threads

**Review by mdemoret-nv — CHANGES_REQUESTED** [review]

File: tests/test_downloader.py [hunk 2]
```diff
@@ -87,29 +88,30 @@ def test_constructor_invalid_dltype(use_env: bool):
         Downloader(**kwargs)
 
 
-@pytest.mark.usefixtures("restore_environ")
+@pytest.mark.reload_modules(morpheus.utils.downloader)
 @pytest.mark.parametrize("dl_method", ["dask", "dask_thread"])
-@mock.patch('dask.config')
+@pytest.mark.usefixtures("reload_modules")
```
Comment (mdemoret-nv) [inline tests/test_downloader.py:93]: Remark: Do we need to use both `@pytest.mark.reload_modules` and `@pytest.mark.usefixtures("reload_modules")` I feel like the call to `usefixtures` could be omitted

File: tests/test_downloader.py [hunk 3]
```diff
@@ -87,29 +88,30 @@ def test_constructor_invalid_dltype(use_env: bool):
...
-def test_get_dask_cluster(mock_dask_cluster: mock.MagicMock, mock_dask_config: mock.MagicMock, dl_method: str):
+def test_get_dask_cluster(mock_dask_cluster: mock.MagicMock, dl_method: str):
     mock_dask_cluster.return_value = mock_dask_cluster
     downloader = Downloader(download_method=dl_method)
     assert downloader.get_dask_cluster() is mock_dask_cluster
 
-    mock_dask_config.set.assert_called_once()
+    # call again then assert that cluster was only created once
+    downloader.get_dask_cluster()
+
```
Comment (mdemoret-nv) [inline tests/test_downloader.py:102]: Suggestion: This test would be better if you called `get_dask_cluster` a second time with a different `Downloader` instance and then assert that only one cluster is created

File: morpheus/utils/downloader.py [hunk 4]
```diff
@@ -124,6 +122,9 @@ def get_dask_client(self):
         """
         import dask.distributed
 
+        # Up the heartbeat interval which can get violated with long download times
+        dask.config.set({"distributed.client.heartbeat": self._dask_heartbeat_interval})
+
```
Comment (mdemoret-nv) [inline morpheus/utils/downloader.py:127]: Question: Why was this moved here? Seems like it would be better to keep it with the dask cluster creation code

File: morpheus/utils/downloader.py [hunk 5]
```diff
@@ -62,12 +62,13 @@ class Downloader:
         The heartbeat interval to use when using dask or dask_thread.
     """
 
+    _dask_cluster = None
```
Comment (mdemoret-nv) [inline morpheus/utils/downloader.py:65]: Important: Add a comment around this describing that this is a cluster shared by all uses of `Downloader` instances.

File: morpheus/utils/downloader.py [hunk 6]
```diff
@@ -96,23 +97,20 @@ def get_dask_cluster(self):
...
 
-            self._dask_cluster = dask_cuda.LocalCUDACluster(n_workers=n_workers, threads_per_worker=threads_per_worker)
+            Downloader._dask_cluster = dask_cuda.LocalCUDACluster(n_workers=n_workers,
+                                                                  threads_per_worker=threads_per_worker)
 
-            logger.debug("Creating dask cluster... Done. Dashboard: %s", self._dask_cluster.dashboard_link)
+            logger.debug("Creating dask cluster... Done. Dashboard: %s", Downloader._dask_cluster.dashboard_link)
 
-        return self._dask_cluster
+        return Downloader._dask_cluster
```
Comment (mdemoret-nv) [inline morpheus/utils/downloader.py:113]: Crucial: This class can be accessed by multiple `Downloader` instances, potentially running in different threads. We need to synchronize access to creating this cluster with a mutex.

**Review by mdemoret-nv — CHANGES_REQUESTED** [review]
Need to synchronize creating the cluster from multiple threads with a mutex.

File: tests/test_downloader.py (hunk 2 above)
Comment (efajardo-nv) [reply tests/test_downloader.py:93]: test fails if `usefixtures` is removed. both are used in our other tests.

File: morpheus/utils/downloader.py (hunk 4 above)
Comment (efajardo-nv) [reply morpheus/utils/downloader.py:127]: moved it to `get_dask_client` because it's a config specific to the dask client. from dask docs:
```
This value is the time between heartbeats The client sends a periodic heartbeat message to the scheduler. If it misses enough of these then the scheduler assumes that it has gone.
```

**Review by mdemoret-nv — APPROVED** [review]
